src/Compilers/VisualBasic/Portable/Symbols/ConstantValueUtils.vb

```diff
@@ -22,44 +22,31 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
     End Class
 
     Friend Module ConstantValueUtils
-        Public Function EvaluateFieldConstant(field As SourceFieldSymbol, equalsValueOrAsNewNodeRef As SyntaxReference, inProgress As SymbolsInProgress(Of FieldSymbol), diagnostics As DiagnosticBag) As EvaluatedConstant
-            Debug.Assert(inProgress IsNot Nothing)
-            Dim value As ConstantValue = Nothing
-            Dim boundValueType As TypeSymbol = Nothing
-            Dim errorField = inProgress.GetStartOfCycleIfAny(field)
-
-            If errorField IsNot Nothing Then
-                diagnostics.Add(ERRID.ERR_CircularEvaluation1,
-                                errorField.Locations(0),
-                                CustomSymbolDisplayFormatter.ShortErrorName(errorField))
-                value = ConstantValue.Bad
-                boundValueType = New ErrorTypeSymbol()
+        Public Function EvaluateFieldConstant(field As SourceFieldSymbol, equalsValueOrAsNewNodeRef As SyntaxReference, dependencies As ConstantFieldsInProgress.Dependencies, diagnostics As DiagnosticBag) As EvaluatedConstant
+#If DEBUG Then
+            Debug.Assert(dependencies IsNot Nothing)
+            Debug.Assert(equalsValueOrAsNewNodeRef IsNot Nothing)
+#End If
+
+            ' Set up a binder for this part of the type.
+            Dim containingModule = field.ContainingSourceType.ContainingSourceModule
+            Dim binder As Binder = BinderBuilder.CreateBinderForType(containingModule, equalsValueOrAsNewNodeRef.SyntaxTree, field.ContainingSourceType)
+            Dim initValueSyntax As VisualBasicSyntaxNode = equalsValueOrAsNewNodeRef.GetVisualBasicSyntax()
+
+            Dim inProgressBinder = New ConstantFieldsInProgressBinder(New ConstantFieldsInProgress(field, dependencies), binder, field)
+            Dim constValue As ConstantValue = Nothing
+            Dim boundValue = BindFieldOrEnumInitializer(inProgressBinder, field, initValueSyntax, diagnostics, constValue)
+
+            Dim boundValueType As TypeSymbol
+
+            ' if an untyped constant gets initialized with a nothing literal, it's type should be System.Object.
+            If boundValue.IsNothingLiteral Then
+                boundValueType = binder.GetSpecialType(SpecialType.System_Object, initValueSyntax, diagnostics)
             Else
-                ' Set up a binder for this part of the type.
-                Dim containingModule = field.ContainingSourceType.ContainingSourceModule
-                Dim binder As Binder = BinderBuilder.CreateBinderForType(containingModule, equalsValueOrAsNewNodeRef.SyntaxTree, field.ContainingSourceType)
-                Dim initValueSyntax As VisualBasicSyntaxNode = equalsValueOrAsNewNodeRef.GetVisualBasicSyntax()
-
-                If initValueSyntax Is Nothing Then
```

Comment:
Was this case just never hit? Trying to understand why we're removing the seeming easy-outs here.

Reply:
>Was this case just never hit? Trying to understand why we're removing the seeming easy-outs here.

I believe so. My understanding is that it is impossible to get null SyntaxNode from a non-null SyntaxReference

---
In reply to: [550661806](https://github.com/dotnet/roslyn/pull/50169#discussion_r550661806) [](ancestors = 550661806)